pkg/auth/oidc_connect.go

```diff
@@ -0,0 +1,294 @@
+	token, err := h.oauth2Config().Exchange(context, code)
+	if err != nil {
+		h.Log.Info(err.Error())
+	}
+
+	rawIDToken, ok := token.Extra("id_token").(string)
+
+	if !ok {
+		resp.Response.StatusCode = http.StatusInternalServerError
+		resp.Allow = false
+		return
+	}
+
+	//Store token.
+	state := store.ConvertToType(stateByte)
+	state.IDToken = rawIDToken
+	state.AccessToken = token.AccessToken
+	state.RefreshToken = token.RefreshToken
+	stateByte = store.ConvertToByte(state)
+	h.Cache.Set(state.OAuthState, stateByte)
+
+	// Set  redirection.
+	resp.Response.StatusCode = http.StatusTemporaryRedirect
+	resp.Response.Header.Add(stateQueryParamName, state.OAuthState)
+
+	// TODO .. may need to rebuild this, in case user could have other query parameters
+	resp.Response.Header.Add("Location", fmt.Sprintf("http://%s?%s=%s", state.RequestPath, stateQueryParamName, state.OAuthState))
+}
+
+func (h *OIDCConnect) isStateValid(ctx context.Context, state *store.OIDCState, provider *oidc.Provider) bool {
+
+	if state == nil {
+		return false
+	}
+
+	verifier := provider.Verifier(&oidc.Config{ClientID: h.OidcConfig.ClientID, SkipIssuerCheck: true})
+
+	// Verify token and signature.
+	idToken, err := verifier.Verify(ctx, state.IDToken)
+	if err != nil {
+		h.Log.Info(fmt.Sprintf("failed to verify ID token: %v", err))
+		return false
+	}
+
+	// Try to claim.
+	var claims json.RawMessage
+	if err := idToken.Claims(&claims); err != nil {
+		h.Log.Info(fmt.Sprintf("error decoding ID token claims: %v", err))
+		return false
+	}
+
+	return true
+}
+
+func (h *OIDCConnect) getStateFromCookie(ctx context.Context, req *Request) (*store.OIDCState, error) {
+
+	// Do you have cookies stored ?
+	cookieVal := req.Request.Header.Get("cookie")
+	var state *store.OIDCState
+	// Check through and get the right cookies
+	if len(cookieVal) > 0 {
+		cookies := strings.Split(cookieVal, ";")
+
+		for _, c := range cookies {
+			c = strings.TrimSpace(c)
+			if strings.HasPrefix(c, oauthTokenName) {
+				cookieJSON := c[len(oauthTokenName)+1:]
+				if len(cookieJSON) > 0 {
+					state = store.ConvertToType([]byte(cookieJSON))
+					return state, nil
+				}
+
+			}
+		}
+	}
+
+	return nil, errors.New("No Cookies available")
+}
+
+func (h *OIDCConnect) initProvider(ctx context.Context) (*oidc.Provider, error) {
+	provider, err := oidc.NewProvider(ctx, h.OidcConfig.IssuerURL)
+	if err != nil {
+		h.Log.Info(fmt.Sprintf("Unable to initialize provider %s", err))
+		return nil, err
+	}
+	return provider, nil
+}
+
+func (h *OIDCConnect) oauth2Config() *oauth2.Config {
+	return &oauth2.Config{
+		ClientID:     h.OidcConfig.ClientID,
+		ClientSecret: h.OidcConfig.ClientSecret,
+		Endpoint:     h.provider.Endpoint(),
+		Scopes:       h.OidcConfig.Scopes,
+		RedirectURL:  h.OidcConfig.RedirectURL + h.OidcConfig.RedirectPath,
+	}
+}
+
+func parseURL(req *Request) *url.URL {
+
+	plainURL, _ := url.QueryUnescape(req.Request.URL.String())
+	u, err := url.Parse(plainURL)
```

Comment:
I don't think it's right to `QueryUnescape` the whole URLs string. We should parse the URL first, then `ParseQuery` on the query string (which does the unescape)

Reply:
some behaviour issue with envoy_grpc, which does not reflect the query parameters. however it does show the full path with QueryString when we call req.Request.URL